### PR TITLE
gh-85934: Use getattr_static when adding mock spec

### PR DESCRIPTION
**Pull request by melwitt**
When async magic method support was added to `unittest.mock.Mock` to address issue [#26467](https://bugs.python.org/issue26467), it introduced a `getattr` call [1] that causes class properties to be called when the class is used as a mock spec.

This caused a problem for a test in my project when running with Python 3.8 where previously the test worked OK with Python 3.6.

The test aims to verify that a class instance is not created if the called code path does not access the class property and thus the class will not create a heavy object unless it's needed (lazy create on access via `@property`).

As of Python 3.8, the `@property` is always called and is called by the mock spec process itself, even though the code path being tested does not access the class `@property`.

Here is a code snippet that illustrates the `@property` calling from the mock spec alone:

```python
class SomethingElse(object):
    def __init__(self):
        self._instance = None

    @property
    def instance(self):
        if not self._instance:
            self._instance = 'object'

...

    def test_property_not_called_with_spec_mock(self):
        obj = SomethingElse()
        self.assertIsNone(obj._instance)
        mock = Mock(spec=obj)
        self.assertIsNone(obj._instance)
```
```bash
$ ./python -m unittest -v unittest.test.testmock.testmock.MockTest.test_property_not_called_with_spec_mock
test_property_not_called_with_spec_mock (unittest.test.testmock.testmock.MockTest) ... FAIL

======================================================================
FAIL: test_property_not_called_with_spec_mock (unittest.test.testmock.testmock.MockTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/cpython/Lib/unittest/test/testmock/testmock.py", line 2173, in test_property_not_called_with_spec_mock
    self.assertIsNone(obj._instance)
AssertionError: 'object' is not None
```

In summary, since commit 77b3b7701a34ecf6316469e05b79bb91de2addfa, class properties are being called when adding mock specs and this can introduce side effects for tests that are verifying state affected by code inside a `@property`.

This replaces the `getattr` call with a `inspect.getattr_static` call to avoid executing class code as part of the mock speccing process.

[1] https://github.com/python/cpython/blob/fb2718720346c8c7a0ad2d7477f20e9a5524ea0c/Lib/unittest/mock.py#L492

https://bugs.python.org/issue41768

* Issue: gh-85934

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@melwitt

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by melwitt** [comment]
Note: I have signed the CLA earlier today, so the records should be updated sometime next week.

File: Lib/unittest/test/testmock/testmock.py [hunk 1]
```diff
@@ -2156,6 +2166,12 @@ def trace(frame, event, arg):  # pragma: no cover
                 obj = mock(spec=Something)
                 self.assertIsInstance(obj, Something)
 
+    def test_property_not_called_with_spec_mock(self):
+        obj = SomethingElse()
+        self.assertIsNone(obj._instance)
```
Comment (terryjreedy) [inline Lib/unittest/test/testmock/testmock.py]:
```suggestion
        self.assertIsNone(obj._instance, msg='before mock')
```

File: Lib/unittest/test/testmock/testmock.py [hunk 2]
```diff
@@ -2156,6 +2166,12 @@ def trace(frame, event, arg):  # pragma: no cover
...
+        mock = Mock(spec=obj)
+        self.assertIsNone(obj._instance)
```
Comment (terryjreedy) [inline Lib/unittest/test/testmock/testmock.py]:
```suggestion
        self.assertIsNone(obj._instance, msg='after mock')
```

File: Lib/unittest/test/testmock/testmock.py [hunk 3]
```diff
@@ -38,6 +38,16 @@ def cmeth(cls, a, b, c, d=None): pass
...
 
+class SomethingElse(object):
+    def __init__(self):
+        self._instance = None
+
+    @property
+    def instance(self):
+        if not self._instance:
+            self._instance = 'object'
+
```
Comment (terryjreedy) [inline Lib/unittest/test/testmock/testmock.py]:
```suggestion
        return self._instance
```
See below for added assert.

**Review by terryjreedy — DISMISSED** [review]
I do not know the mock specification well enough to be sure of what it should do, except that modifying the underlying object is likely wrong, and that the mock should actually mock it.  So the code patch is likely correct and an additional test seems needed.

File: Lib/unittest/test/testmock/testmock.py (hunk 3 above)
Comment (melwitt) [reply Lib/unittest/test/testmock/testmock.py]: Hi, thank you for the review! I don't see an added assert that makes use of this part of the change, am I missing something?

**Comment by melwitt** [comment]
> I do not know the mock specification well enough to be sure of what it should do, except that modifying the underlying object is likely wrong, and that the mock should actually mock it. So the code patch is likely correct and an additional test seems needed.

Yes, when you create a mock and provide a spec to it (a class or object) the mock will be created to only allow the interface of the class spec. That way, arbitrary calls upon your mock with invalid methods or attributes will fail in your test.

In order to create the spec, it needs to inspect the class or object but ideally it should not call any methods on the object or change its state (IMHO) while doing so. Prior to Python 3.8 it did not, but the more recent async mock support checks whether methods are coroutine functions and its use of getattr is now causing object properties to be called.

Here's a section in the docs that explains speccing as part of an explanation of auto-speccing:

https://docs.python.org/3/library/unittest.mock.html#autospeccing

File: Lib/unittest/test/testmock/testmock.py [hunk 4]
```diff
@@ -2156,6 +2166,12 @@ def trace(frame, event, arg):  # pragma: no cover
                 obj = mock(spec=Something)
                 self.assertIsInstance(obj, Something)
 
+    def test_property_not_called_with_spec_mock(self):
+        obj = SomethingElse()
+        self.assertIsNone(obj._instance)
+        mock = Mock(spec=obj)
+        self.assertIsNone(obj._instance)
+
```
Comment (terryjreedy) [inline Lib/unittest/test/testmock/testmock.py]:
```suggestion
        self.assertEqual(mock.instance, 'instance')
```

File: Lib/unittest/test/testmock/testmock.py (hunk 3 above)
Comment (terryjreedy) [reply Lib/unittest/test/testmock/testmock.py]: The suggestion that I wrote but did not actually add until now.

Comment (melwitt) [reply Lib/unittest/test/testmock/testmock.py]: A-ha, thanks! The new assert makes sense though I'm noticing I think this new line should be unindented one level. I don't see a way I can edit a suggestion when I add it to a batch.

Comment (melwitt) [reply Lib/unittest/test/testmock/testmock.py]: Sigh, sorry, my mistake. I don't know what is wrong with my eyes today.

**Comment by melwitt** [comment]
Gah, I have really messed this up by failing at using the UI here. Apologies, working on fixing it.

**Comment by melwitt** [comment]
Windows test suite failure appears unrelated.

**Comment by melwitt** [comment]
I have made the requested changes; please review again

**Comment by terryjreedy** [comment]
I could not download logs, so when one can, better to briefly summarize the failure in a comment.  I triggered the minimum retest possible.

**Review by terryjreedy — COMMENTED** [review]
Changes match those requested.  I am hoping someone else reviews this, but I will read spec reference if not.

**Comment by melwitt** [comment]
> I could not download logs, so when one can, better to briefly summarize the failure in a comment. Ah, you did, but I did not see on email. I triggered the minimum retest possible.

Sorry about that, I'm a bit new to the github PR workflow for code review and am missing some things.

After I wrote "appears unrelated" and posted it, I realized I should have included a paste of the error and a link to the log so I made an edit afterward. Thanks for your patience.

**Comment by willingc** [comment]
Hi @melwitt, Thanks for working on this PR. If you would like to move this PR forward, please resolve the current conflicts. Thanks! 🐍

**Comment by terryjreedy** [comment]
The conflict was that #100252 removed the 3-line for-if loop that this PR proposes to fix.

**Comment by carljm** [comment]
@terryjreedy My PR didn't remove this loop, but it indented it one level deeper and moved it down about ten lines. The change from `getattr` to `inspect.getattr_static` could still apply to the moved loop, but we definitely shouldn't reintroduce a second copy of the loop, as the current PR does. 

`getattr_static` will be slower than `getattr`, but that shouldn't block this change. I would be a little worried about the backward-compatibility impact of changing this behavior, although I do think that not executing properties is probably better behavior if this were being designed from scratch.

**Review by carljm — CHANGES_REQUESTED** [review]
If we do this, the change from `getattr` to `getattr_static` should be made in the existing loop on line 533; we shouldn't introduce a new copy of this loop.

**Review by carljm — APPROVED** [review]
The two possible concerns here are that `getattr_static` is slower than `getattr`, and backward-compatibility. But I think mock speccing executing properties is a bug, and fixing that bug outweighs those two concerns. Also, @AlexWaygood made `getattr_static` faster in 3.12. And this behavior of mock spec executing properties was introduced relatively recently (with AsyncMock); that change was a back-compat break from previous longstanding behavior, and this is the fix for it.

Because this is a bug, I think the fix should also be backported.

File: Misc/NEWS.d/next/Library/2020-09-16-16-53-06.bpo-41768.8_fWkC.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+unittest.mock speccing no longer calls class properties
```
Comment (arhadthedev) [inline Misc/NEWS.d/next/Library/2020-09-16-16-53-06.bpo-41768.8_fWkC.rst]:
```suggestion
:mod:`unittest.mock` speccing no longer calls class properties.
Patch by Melanie Witt.
```

**Review by carljm — CHANGES_REQUESTED** [review]
Currently async methods decorated with `@classmethod` or `@staticmethod` are detected as async attributes. This PR causes them not to be anymore. I think we need to preserve this behavior (and add tests for it.) We'll have to follow `__wrapped__` to the decorated method, and test it with `iscoroutinefunction`. We can do this using `inspect.unwrap`.

**Comment by melwitt** [comment]
I have made the requested changes; please review again (per bedevere-bot)

**Comment by melwitt** [comment]
AFAICT the `test__xxsubinterpreters` failure in the [Hypothesis Tests on Ubuntu](https://github.com/python/cpython/actions/runs/4988297478/jobs/8930880863#step:19:56) job is unrelated:
 
```
0:00:18 load avg: 2.10 [ 18/427/1] test__xxsubinterpreters crashed (Exit code -6)

Fatal Python error: drop_gil: PyMUTEX_LOCK(gil->switch_mutex) failed
Python runtime state: initialized

Thread 0x00007fdc1efb44c0 (most recent call first):
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/test__xxsubinterpreters.py", line 80 in clean_up_interpreters
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/test__xxsubinterpreters.py", line 88 in tearDown
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/case.py", line [59](https://github.com/python/cpython/actions/runs/4988297478/jobs/8930880863?pr=22209#step:19:60)4 in _callTearDown
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/case.py", line 637 in run
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/case.py", line 690 in __call__
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/suite.py", line 122 in run
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/suite.py", line 84 in __call__
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/suite.py", line 122 in run
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/suite.py", line 84 in __call__
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/suite.py", line 122 in run
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/suite.py", line 84 in __call__
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/runner.py", line 240 in run
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/support/__init__.py", line 1115 in _run_suite
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/support/__init__.py", line 1241 in run_unittest
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/runtest.py", line 294 in _test_module
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/runtest.py", line 330 in _runtest_inner2
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/runtest.py", line 373 in _runtest_inner
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/runtest.py", line 232 in _runtest
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/runtest.py", line 278 in runtest
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/runtest_mp.py", line 98 in run_tests_worker
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/main.py", line 759 in _main
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/main.py", line 738 in main
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/main.py", line 802 in main
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/regrtest.py", line 43 in _main
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/regrtest.py", line 47 in <module>
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/runpy.py", line 88 in _run_code
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/runpy.py", line 198 in _run_module_as_main
Fatal Python error: Aborted

Thread 0x00007fdc1efb44c0 (most recent call first):
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/test__xxsubinterpreters.py", line 80 in clean_up_interpreters
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/test__xxsubinterpreters.py", line 88 in tearDown
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/case.py", line 594 in _callTearDown
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/case.py", line [63](https://github.com/python/cpython/actions/runs/4988297478/jobs/8930880863?pr=22209#step:19:64)7 in run
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/case.py", line 690 in __call__
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/suite.py", line 122 in run
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/suite.py", line 84 in __call__
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/suite.py", line 122 in run
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/suite.py", line 84 in __call__
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/suite.py", line 122 in run
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/suite.py", line 84 in __call__
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/unittest/runner.py", line 240 in run
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/support/__init__.py", line 1115 in _run_suite
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/support/__init__.py", line 1241 in run_unittest
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/runtest.py", line 294 in _test_module
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/runtest.py", line 330 in _runtest_inner2
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/runtest.py", line 3[73](https://github.com/python/cpython/actions/runs/4988297478/jobs/8930880863?pr=22209#step:19:74) in _runtest_inner
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/runtest.py", line 232 in _runtest
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/runtest.py", line 278 in runtest
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/runtest_mp.py", line 98 in run_tests_worker
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/main.py", line [75](https://github.com/python/cpython/actions/runs/4988297478/jobs/8930880863?pr=22209#step:19:76)9 in _main
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/main.py", line 738 in main
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/libregrtest/main.py", line [80](https://github.com/python/cpython/actions/runs/4988297478/jobs/8930880863?pr=22209#step:19:81)2 in main
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/regrtest.py", line 43 in _main
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/regrtest.py", line 47 in <module>
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/runpy.py", line 88 in _run_code
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/runpy.py", line 198 in _run_module_as_main

Extension modules: _testcapi, _xxsubinterpreters (total: 2)
```

**Review by carljm — APPROVED** [review]

**Comment by carljm** [comment]
Thanks @melwitt for the patch!
